cohortextractor2/concepts/constraints.py

```diff
@@ -0,0 +1,42 @@
+class BaseConstraint:
+    """
+    Base class for Constraints to be enforced on Columns defined in a Contract
+    """
+
+    def validate(self, database):
+        """Validate against data"""
+        raise NotImplementedError
+
+
+class ChoiceConstraint(BaseConstraint):
+    """
+    Defines a constraint on a choices Column to ensure data matches the allowed options
+    """
+
+    def __init__(self, *choices):
+        self.choices = choices
+
+    def validate(self, backend_table, column):
+        """Validate against data"""
+        # TODO
+
+
+class DateConstraint(BaseConstraint):
+    """
+    Defines a constraint on a date Column to ensure it conforms to required date format(s)
+    """
+
+    base_format = "%Y-%m-%d"
+
+    def __init__(self, match_format: str):
+        """
+        match_format: str.  A constraint format in the form '%Y-%m-%d' to compare with the base format.
+        e.g. '%Y-%m-01' ensures that each date in the backend column repesents the 1st of a month
+        """
```

Comment:
I'm doubtful that the generality you're providing for here will _really_ prove to be useful. Maybe better to have a date-is-first-of-month constraint, which might also make generating documentation for it easier, them generalize later if there are other similar constraints that we need to support.

Reply:
OK, sure, that makes sense